doc/design/katib_on_sqlflow.md

```diff
@@ -1,70 +1,91 @@
+Argo takes workflows in the form of YAML files, and it is error-prone to write such YAML files manually.  So, we created [Couler](/python/couler/README.md) as an intermediate programmatic representation of workflows.
 
-Then it needs to invoke different Couler APIs according to input SQL queries. In Katib case, Couler codegen needs to check if the input SQL query is to create a Katib job. If it is, the codegen should generate codes to invoke a Couler Katib API, like `couler.run_katib(...)`
+We need to develop a new codegen, `codegen_couler.go`, for SQLFlow.  `codegen_couler.go` converts the parsed SQL program, a.k.a., the [intermediate representation](/pkg/sql/ir), or IR, into a Couler program.
 
-## Design for `couler.run_katib(...)`
+### The Integration via Couler
 
-Considering Katib itself supports multiple models and frameworks, and more may come in the future, we use a unique API to create Katib steps in Couler. The API is as following:
+SQLFlow parses each SQL program into an IR, which is a list of statement IRs.  The `codegen_couler.go` converts the IR into a Couler program.   We need to add a Couler functions `couler.katib.train` for the calling by the generated Couler program.
 
-`couler.run_katib("model_or_framework"=None, katib_params={}, model_or_framework_params={})`
+Consider the following example program.
 
-In this API, it includes three arguments:
-- `model_or_framework`: string, indicates the model (e.g., XGBoost) or frameworks (e.g., tf or pytorch) used in model training on Katib.
-- `katib_params`: dict, configures Katib jobs. For example, users can specify max trials in model training, like `katib_params= { "max_trial_count": 10}`  
-- `model_or_framework_params`: dict, configures model or framework given in `model_or_framework`. The above XGBoost example will be: `model_or_framework_params= { "booster": "gbtree", "objective": "multi:softmax", "eta": 1, "num_round": [20, 100], max_depth: [] }`
+```sql
+SELECT * FROM a, b WHERE a.id = b.id INTO c;
+SELECT * FROM c TO TRAIN model_def WITH objective=multi:softmax, eta=1 LABEL class INTO my_xgb_model;
+```
+
+The `codegen_couler.go` might generate the following Couler program.
+
+```python
+couler.maxcompute.run("""SELECT * FROM a, b WHERE a.id = b.id INTO c;
+                         SELECT * FROM c INTO temp""")
+couler.maxcompute.export(table="temp", file="/hdfs/temp")
+couler.katib.train(model_def, data="/hdfs/temp")
+```
+
+## `couler.katib.train(...)`
+
+Considering Katib itself supports multiple models and frameworks, and more may come in the future, we introduce the following Couler function.
+
+```python
+def couler.katib.train(model_def=None, hyperparameters={})
```

Comment:
Can you show more details about the khatib train step, such as:
1. where does the model_def come from in SQLFlow?
1. In this example, `hyperparameters ` is empty, what and how does this argument to use in SQLFlow?
1. Do we need `couler.katib.predict` ?

Reply:
Sure.
1.1 As @wangkuiyi suggested, programmers need not specify to use katib explicitly. Instead, sqlflow will check if programmers have input all necessary model parameters. If it is, sqlflow invokes submitter to train model (e.g., xgboost submiitter); otherwise, sqlflow uses katib via couler.
1.2  As the xgboost example above, the model_def can be something like `xgboost:gbtree`.
2. `hyperparameters={"booster": "gbtree",  "objective": "multi:softmax", "eta": 1.0, "max_round": [20, 100], "max_depth: []" }`.
3. No. Katib is only used for model tuning.